File: 0061_rotate_list/Solution.cs
// LeetCode 0061 - Rotate List
// https://leetcode.com/problems/rotate-list/

public class ListNode {
    public int val;
    public ListNode next;
    public ListNode(int val = 0, ListNode next = null) {
        this.val = val;
        this.next = next;
    }
}

public class Solution {
    public ListNode RotateRight(ListNode head, int k) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode tail = head;
        int length = 1;
        while (tail.next != null) {
            tail = tail.next;
            length++;
        }

        tail.next = head;
        k %= length;
        if (k == 0) {
            tail.next = null;
            return head;
        }

        int steps = length - k;
        ListNode newTail = head;
        for (int i = 0; i < steps - 1; i++) {
            newTail = newTail.next;
        }

        ListNode newHead = newTail.next;
        newTail.next = null;
        return newHead;
    }
}

File: 0061_rotate_list/Solution.scala
// LeetCode 0061 - Rotate List
// https://leetcode.com/problems/rotate-list/

class ListNode(var _x: Int = 0) {
  var next: ListNode = null
  var x: Int = _x
}

object Solution {
  def rotateRight(head: ListNode, k: Int): ListNode = {
    if (head == null || head.next == null) {
      return head
    }

    var tail = head
    var length = 1
    while (tail.next != null) {
      tail = tail.next
      length += 1
    }

    tail.next = head
    var remaining = k % length
    if (remaining == 0) {
      tail.next = null
      return head
    }

    val steps = length - remaining
    var newTail = head
    var i = 0
    while (i < steps - 1) {
      newTail = newTail.next
      i += 1
    }

    val newHead = newTail.next
    newTail.next = null
    newHead
  }
}
